### PR TITLE
Intra-graph reordering pass on Inductor scheduler IR (based on #100762)

### PR DESCRIPTION
**Pull request by yf225**
This PR implements intra-graph communication reordering pass on Inductor scheduler IR, based on Horace's previous PR #100762.

Main algorithm:
1. Greedily moves waits as late as possible (i.e. until we reach a use)
2. Greedily moves comms as early as possible (i.e. until we reach an input)
3. Move computes following simple heuristics to improve overlap.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler @ngimel

### CODE REVIEW

File: torch/_inductor/config.py [hunk 1]
```diff
@@ -74,6 +74,9 @@
 # If not specified, a temp directory will be created under the default caching path
 aot_inductor_output_path = ""
 
+# enable reordering pass for increasing overlap between compute and communication
+reorder_for_compute_comm_overlap = False
```
Comment (wanchaol) [inline torch/_inductor/config.py:78]: curious how we can turn on this config when using torch.compile?

**Review by wanchaol — COMMENTED** [review]
Wondering if could we add some tests to verify the re-ordering works properly?

File: torch/_inductor/analysis.py [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+from . import ir
+
+def get_runtime_of_snode(snode: "BaseSchedulerNode"):
+    """
+    Gets the runtime of Scheduler node. Currently somewhat of a placeholder, todo to be replaced by more sophisticated approaches.
+
+    TODO(yf225): potentially replace with a more sophisticated cost model (e.g. analytical)
+    """
+    # print(f"get_runtime_of_snode: snode.node: {snode.node}")
+    if isinstance(snode.node, ir.AllReduce):
```
Comment (wanchaol) [inline torch/_inductor/analysis.py:12]: another q: do you plan to add `allgather` and `reduce_scatter` in follow up PRs? asking bc in training ppl mostly use sequence parallel and it would be great if we can add these two :)

Comment (yf225) [reply torch/_inductor/analysis.py:12]: all other collectives will go to the `ir.CollectiveKernel` branch, also we desperately need a better cost model for this..

File: torch/_inductor/analysis.py [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+    Gets the runtime of Scheduler node. Currently somewhat of a placeholder.
+
+    TODO: replace with a more sophisticated cost model (e.g. analytical)
+    """
+    # Collective kernels
+    if isinstance(snode.node, ir.CollectiveKernel):
+        if isinstance(snode.node, ir.AllReduce):
+            return 5
+        else:
+            return 15
```
Comment (wconstab) [inline torch/_inductor/analysis.py:16]: how come other collectives are given this higher cost?

File: torch/_inductor/comms.py [hunk 4]
```diff
@@ -0,0 +1,277 @@
...
+        for node in cur_nodes:
+            for inp in node.node_users:
+                if inp not in descendants:
+                    descendants.add(inp)
+                    new_nodes.append(inp)
+        cur_nodes = new_nodes
+    return descendants
+
+
+def decide_global_ordering_of_comms(nodes: List["scheduler.BaseSchedulerNode"]):
```
Comment (wconstab) [inline torch/_inductor/comms.py:102]: a couple questions on global ordering

1) is the approach that we are not allowed to change the relative order of comm ops from the original program? 

If so, that's pretty safe, but have you reasoned that it's good enough?  (if we consider TP and DP cases, will we need to move a DP comm past a TP comm?)

File: torch/_inductor/comms.py [hunk 5]
```diff
@@ -0,0 +1,277 @@
...
+
+import torch.fx as fx
+
+from . import ir, scheduler, config
+from .analysis import get_snode_runtime
+from .dependencies import WeakDep
+
+
+# Used to ensure that iterating over a set is deterministic
+def tuple_sorted(x):
```
Comment (Chillee) [inline torch/_inductor/comms.py:14]: Worth putting in `torch/_inductor/utils.py`

File: torch/_inductor/comms.py [hunk 6]
```diff
@@ -0,0 +1,277 @@
...
+    """
+    Decides a global ordering of all compute and communication nodes. Assumes that we already have a global ordering of communication nodes.
+    Overall scheduling procedure is:
+        Step 1: Given that we've currently scheduled comm N, we now schedule all compute nodes that are required for comm N + 1 but do not depend on comm N, to run at the same time with comm N.
+        Step 2: If all those compute nodes are sufficient to overlap comm N, we're done. Otherwise, we now need to look elsewhere to find compute that overlaps with comm N. We prioritize compute nodes that are needed sooner.
+        Step 3: We schedule the compute nodes dependent on comm N and required for comm N + 1.
+        Step 4: We schedule comm N + 1.
+        Repeat this for subsequent comm nodes.
+    """
+    result = []
```
Comment (Chillee) [inline torch/_inductor/comms.py:134]: Rename to something like `final_order`.

File: torch/_inductor/comms.py [hunk 7]
```diff
@@ -0,0 +1,277 @@
...
+                    schedule_node(node)
+                    all_nodes.remove(node)
+                    progress = True
+            if not progress:
+                raise Exception("Unable to find a free node (indeg == 0). This is an impossible state to reach. Please report a bug to PyTorch.")
+
+    # First, schedule all compute nodes that are required by first comm node,
+    # as well as the first comm node itself.
+    schedule_nodes(
+        list(comm_ancestors[comm_nodes[0]]) + [comm_nodes[0]],
```
Comment (Chillee) [inline torch/_inductor/comms.py:191]: Do we need to check if there are any `comm_nodes` and if not, return early?

File: test/distributed/test_compute_comm_reordering.py [hunk 8]
```diff
@@ -0,0 +1,267 @@
...
+            e = d + ar
+            return (e,)
+
+        with _dynamo_dist_per_rank_init(self.rank, self.world_size):
+            inputs = torch.ones(4, 4, dtype=torch.float, device="cuda") + self.rank
+            compiled = torch.compile(func)
+            code = run_and_get_triton_code(compiled, inputs, **self.get_world_trs())
+            # NOTE: notice that `_wait_tensor` is delayed until right before first use
+            FileCheck() \
+                .check("buf0 = empty_strided((4, 4), (4, 1), device='cuda', dtype=torch.float32)") \
```
Comment (Chillee) [inline test/distributed/test_compute_comm_reordering.py:87]: I'm not a big fan of these FileCheck tests, matching against the entire generated output. I think there's too much churn from random things that get changed, and updating these tests will be quite painful.

One option is to use `assertExpectedEqual`, which at least provides a more automated way of optimizing it.

Another thing we can do is after our scheduling, we can write a simple simulator based off of our nodes' estimated runtime to compute the total runtime and things like "exposed comms".

Comment (Chillee) [reply test/distributed/test_compute_comm_reordering.py:87]: That would also allow us to potentially run these tests without actually needing multiple GPUs, which would be a huge win I think.

**Review by Chillee — COMMENTED** [review]
Mostly looks good to me I think.

Main comment are on the tests.

File: torch/_inductor/comms.py [hunk 9]
```diff
@@ -0,0 +1,277 @@
...
+            new_result.append(snode)
+    for snode in tuple_sorted(cur_waits):
+        new_result.append(snode)
+    return new_result
+
+
+def raise_comms(result: List["scheduler.BaseSchedulerNode"]) -> List["scheduler.BaseSchedulerNode"]:
+    """
+    Greedily moves comms as early as possible (i.e. until we reach an input).
+    Optimal in terms of communication overlap.
```
[the remark this answers is not in the document]
Comment (wconstab) [reply torch/_inductor/comms.py:53]: yea, it seems like we'll have to either do a special pass for FSDP (meh..) or we'll want to redo this pass with memory also being considered so we handle the FSDP case in a general way

[the remark this answers is not in the document]
Comment (yf225) [reply torch/_inductor/comms.py:37]: added comment to document this

File: torch/_inductor/analysis.py [hunk 10]
```diff
@@ -0,0 +1,23 @@
...
+from . import ir
+
+def get_snode_runtime(snode: "BaseSchedulerNode"):
+    """
+    Gets the runtime of Scheduler node. Currently somewhat of a placeholder.
+
+    TODO: replace with a more sophisticated cost model (e.g. analytical)
+    """
+    # Collective kernels
+    if isinstance(snode.node, ir.CollectiveKernel):
```
[the remark this answers is not in the document]
Comment (yf225) [reply torch/_inductor/analysis.py]: added runtime estimation for collective ops, based on https://github.com/NVIDIA/nccl/blob/master/src/graph/tuning.cc

**Review by Chillee — APPROVED** [review]
Looks good to me I think.

File: torch/_inductor/comm_analysis.py [hunk 11]
```diff
@@ -0,0 +1,179 @@
...
+    """
+    Returns estimated NCCL collective runtime in nanoseconds (ns).
+
+    The following heuristics are copied from https://github.com/NVIDIA/nccl/blob/master/src/graph/tuning.cc.
+    We aim to estimate the runtime as accurately as possible.
+
+    Assumptions:
+    - only ring algorithm (NCCL_ALGO_RING) is used
+    - only Low-Latency protocol (NCCL_PROTO_LL) is used, i.e. Simple or LL128 is not used
+    - only A100 gpu
```
Comment (wanchaol) [inline torch/_inductor/comm_analysis.py:90]: This looks great! 

As a follow up we might can see how to get those numbers from some NCCL APIs directly instead of hard code them, as it might be fixed for one hardware

File: torch/_inductor/comm_analysis.py [hunk 12]
```diff
@@ -0,0 +1,179 @@
...
+    - inter-node is only InfiniBand
+    - collective is one of: allreduce, reducescatter, allgather
+    """
+    tensor_numel = V.graph.sizevars.size_hint(sympy_product(snode.node.layout.size))
+    tensor_dtype = snode.node.layout.dtype
+    tensor_storage_size_bytes = tensor_numel * get_dtype_size(tensor_dtype)
+    # Convert bytes to GB
+    tensor_storage_size_GB = tensor_storage_size_bytes / 1024 / 1024 / 1024
+
+    # TODO: is there a way to get accurate "gpus per node" and "# nodes" info?
```
Comment (wanchaol) [inline torch/_inductor/comm_analysis.py:102]: yeah from the current process group API probably no, we might can see how to get those infos from NCCL. one thing I tried to mimic was to use the default world size / # gpus per device, but this might not be super reliable https://github.com/pytorch/pytorch/pull/109145/files#diff-a2ab7823568ba9f2a2818ec629dc9208df45953bb8687ed9b7791753ef65e9e7R93

File: test/distributed/test_compute_comm_reordering.py [hunk 13]
```diff
@@ -0,0 +1,250 @@
...
+            return 5
+        # All other kernels
+        return 1
+
+    @unittest.skipIf(not has_triton(), "Inductor+gpu needs triton and recent GPU arch")
+    @skip_if_lt_x_gpu(2)
+    @patch.object(torch._inductor.config, "allow_buffer_reuse", True)
+    # TODO: somehow inductor bg compile threads are causing hangs at exit with distributed work dtor
+    @patch.object(torch._inductor.config, "compile_threads", 1)
+    @patch.object(torch._inductor.config, "reorder_for_compute_comm_overlap", True)
```
Comment (wanchaol) [inline test/distributed/test_compute_comm_reordering.py:211]: for usage, we can simply do torch._inductor.config.reorder_for_compute_comm_overlap = True and no need to patch the passes right?

**Review by wanchaol — APPROVED** [review]
lgtm

File: torch/_inductor/comm_analysis.py [hunk 14]
```diff
@@ -0,0 +1,179 @@
...
+        return ncclFuncAllReduce
+    elif isinstance(snode.node, (ir.AllGatherIntoTensor, ir.AllGatherIntoTensorCoalesced)):
+        return ncclFuncAllGather
+    elif isinstance(snode.node, (ir.ReduceScatterTensor, ir.ReduceScatterTensorCoalesced)):
+        return ncclFuncReduceScatter
+    else:
+        raise Exception(f"Unsupported collective type: {snode.node}")
+
+
+def estimate_nccl_collective_runtime(snode: "scheduler.BaseSchedulerNode") -> float:
```
[the remark this answers is not in the document]
Comment (yf225) [reply torch/_inductor/comm_analysis.py:80]: Tested it for allreduce on 2-GPU:

actual vs. predicted
[4,4], 2-gpu: 7us vs 7.8us
[16,16], 2-gpu: 7us vs 7.8us
[128, 128], 2-gpu: 8.5us vs. 10us
[1024, 1024], 2-gpu: 300us vs 50us
[2048, 2048], 2-gpu: 284us vs 186us
[10240, 10240], 2-gpu: 6146us vs 4462us

it's not too bad (maybe like +/- 30% accuracy)

File: torch/_inductor/comm_analysis.py (hunk 11 above)
Comment (yf225) [reply torch/_inductor/comm_analysis.py:90]: agreed - the original NCCL code has V100 and H100 numbers too, we can expand this in follow-up, or find a way to have an NCCL API for it

File: test/distributed/test_compute_comm_reordering.py (hunk 13 above)
Comment (yf225) [reply test/distributed/test_compute_comm_reordering.py:211]: yes that's right

File: test/distributed/test_compute_comm_reordering.py (hunk 8 above)
Comment (wconstab) [reply test/distributed/test_compute_comm_reordering.py]: > That would also allow us to potentially run these tests without actually needing multiple GPUs, which would be a huge win I think.

couldn't we run a test with collectives in it on a single gpu today? the collective should still work at all layers even if its essentially a no-op.

> I'm not a big fan of these FileCheck tests

its definitely fragile to match too much of the generated code.  especially variable names.. but unless we actually build some more infra for validation of certain properties of IR its the best we have.  

What does assertExpectedEqual do that is helpful here?

File: torch/_inductor/comm_analysis.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (wconstab) [reply torch/_inductor/comm_analysis.py:80]: well.. more than 2 gpu is also important, i'm guessing the variance only goes higher in larger scale.

more important than how accurate the model is, how sensitive is the optimization to wrong model output?

[the remark this answers is not in the document]
Comment (wconstab) [reply torch/_inductor/comm_analysis.py:80]: upon looking more closely, i am not in favor of landing this cost model.  It seems too specialized on the small-scale and currently used hw setup.  IF its important to have this much detail in a model, that implies its important for it to be accurate at large scale and across different system types.  IF its not important for this to be accurate, just replace it with a simpler heuristic.

File: test/distributed/test_compute_comm_reordering.py (hunk 8 above)
Comment (Chillee) [reply test/distributed/test_compute_comm_reordering.py]: > What does assertExpectedEqual do that is helpful here?

It allows us to automatically update all test files that have jitter in the output.

> couldn't we run a test with collectives in it on a single gpu today? the collective should still work at all layers even if its essentially a no-op.

Does this work? Even through Inductor?

Comment (wconstab) [reply test/distributed/test_compute_comm_reordering.py]: > Does this work? Even through Inductor? Well regardless, we'd still want property-based testing.

well.. it definitely works in eager. i don't see why it wouldn't work in inductor but i didn't try it

Comment (Chillee) [reply test/distributed/test_compute_comm_reordering.py]: I'm wondering whether it gets caught and dispatched to a clone somewhere.

File: torch/_inductor/comm_analysis.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (yf225) [reply torch/_inductor/comm_analysis.py]: I believe accurate is still the best, but it's not super important currently because a somewhat (not 100%) accurate model is still better than a dummy model.

I'd actually argue that a simpler heuristic will not be much simpler in terms of the code, because we still need to consider both latency and bandwidth terms, and the computation will be quite similar to current code. E.g. for bandwidth calculation which is important for large-scale case, there is a fixed formula for calculating steps used in ring algorithm, and we will need to use the same formula in simpler heuristics. And for small-scale, latency dominates and we don't know the empirical latency numbers unless we copy from existing NCCL code.

To make this more generic, I can add V100 and H100 heuristics to the code. I believe NVLink for intra-node is mostly safe to assume, and I can add a user-facing config for intra-node interconnect bandwidth (e.g. if user is using AWS EFA instead of InfiniBand).

[the remark this answers is not in the document]
Comment (yf225) [reply torch/_inductor/comm_analysis.py]: Synced offline, I will add:
1. V100 and H100 heuristics
2. user-facing configs for intra-node and inter-node interconnect bandwidth

File: torch/_inductor/comm_analysis.py [hunk 15]
```diff
@@ -0,0 +1,232 @@
...
+            87.7,
+            22.5,  # avg of ring & tree
+            19.0,
+        ],
+    ]
+)
+
+
+def get_gpu_type() -> NVIDIA_GPU_TYPE:
+    gpu_info = torch.utils.collect_env.get_gpu_info(torch.utils.collect_env.run)
```
Comment (int3) [inline torch/_inductor/comm_analysis.py:104]: I don't think `torch.utils.collect_env` works -- it doesn't typecheck and the module lookup doesn't work at runtime either. I guess we probably need to add an `import collect_env` in `__init__.py`

Comment (yf225) [reply torch/_inductor/comm_analysis.py:104]: thanks for the catch and you are right. Fix coming..

**Comment by zejun-chen** [comment]
Hi, @yf225 

Wonderful work, may i ask a question about this pass?
Usually the users may put the collective ops to a new cuda Stream for overlapping with the compute ops in the current Stream, but it looks like the Stream ops(for example, `torch.cuda.set_stream`) will be ignored by the `torch.compile`.
Here the inductor will schedule the collective ops on the graph to another Stream? or both collective ops and compute ops will execute on the same Stream?

Thank you.
